crates/ruff_python_parser/src/string.rs

```diff
@@ -535,68 +227,95 @@ impl<'a> StringParser<'a> {
+    };
+    StringParser::new(source.as_str(), kind, start)
+        .parse_fstring_middle()
+        .map(Some)
 }
 
-pub(crate) fn parse_strings(
-    values: Vec<(TextSize, (String, StringKind, bool), TextSize)>,
-) -> Result<Expr, LexicalError> {
-    // Preserve the initial location and kind.
-    let initial_start = values[0].0;
-    let last_end = values.last().unwrap().2;
-    let initial_kind = (values[0].1 .1 == StringKind::Unicode).then(|| "u".to_owned());
-    let has_fstring = values
+/// Concatenate a list of string expressions into a single string expression.
+///
+/// This is mainly used for implicit string concatenation and the possible
+/// expression values are strings, bytes, and f-strings.
+pub(crate) fn concatenate_strings(
+    strings: Vec<ParenthesizedExpr>,
+    range: TextRange,
+) -> Result<ParenthesizedExpr, LexicalError> {
+    #[cfg(debug_assertions)]
+    debug_assert!(!strings.is_empty());
+
+    let has_fstring = strings
         .iter()
-        .any(|(_, (_, kind, ..), _)| kind.is_any_fstring());
-    let num_bytes = values
+        .any(|parenthesized_expr| parenthesized_expr.expr.is_f_string_expr());
+    let num_bytes = strings
         .iter()
-        .filter(|(_, (_, kind, ..), _)| kind.is_any_bytes())
+        .filter(|parenthesized_expr| {
+            matches!(
+                parenthesized_expr.expr,
+                Expr::Constant(ast::ExprConstant {
+                    value: Constant::Bytes(_),
+                    ..
+                })
+            )
+        })
         .count();
     let has_bytes = num_bytes > 0;
-    let implicit_concatenated = values.len() > 1;
+    let implicit_concatenated = strings.len() > 1;
 
-    if has_bytes && num_bytes < values.len() {
+    if has_bytes && num_bytes < strings.len() {
         return Err(LexicalError {
             error: LexicalErrorType::OtherError(
                 "cannot mix bytes and nonbytes literals".to_owned(),
             ),
-            location: initial_start,
+            location: range.start(),
         });
     }
 
     if has_bytes {
         let mut content: Vec<u8> = vec![];
-        for (start, (source, kind, triple_quoted), _) in values {
-            for value in parse_string(&source, kind, triple_quoted, start)? {
-                match value {
-                    Expr::Constant(ast::ExprConstant {
-                        value: Constant::Bytes(BytesConstant { value, .. }),
-                        ..
-                    }) => content.extend(value),
-                    _ => unreachable!("Unexpected non-bytes expression."),
-                }
+        for parenthesized_expr in strings {
+            match parenthesized_expr.expr {
+                Expr::Constant(ast::ExprConstant {
+                    value: Constant::Bytes(BytesConstant { value, .. }),
+                    ..
+                }) => content.extend(value),
+                _ => unreachable!("Unexpected non-bytes expression."),
             }
         }
```

Comment:
I wonder if we should add a fast path (for the whole method) that handles the most common case of a single string:

* No need to test if it is a mixed byte / non-byte concatenation
* No need to copy the value (value.into_bytes() or use the original string)